src/aff_expr.jl

```diff
@@ -16,12 +16,16 @@
 #############################################################################
 
 # Utilities for OrderedDict
-function _add_or_set!(dict::OrderedDict{K,V}, k::K, v::V) where {K,V}
-    # TODO: This unnecessarily requires two lookups for k.
+function _add_or_set!(expr_terms::OrderedDict{K,V}, k::K, v::V) where {K,V}
     # TODO: Decide if we want to drop zeros here after understanding the
     # performance implications.
-    dict[k] = get!(dict, k, zero(V)) + v
-    return dict
+    v0 = get!(expr_terms, k, v)
+    # This makes sure only one hash lookup is need when inserting a value.
+    if v0 !== v
```

Comment:
This doesn't work when `expr_terms[k]` is equal to `v`.

Reply:
Good catch. Glad we have tests :)

Reply:
So it isn't an unnecessary double look-up. You need one to get the current value (if it exists), and the second to set the new value. I don't see a way to get around this.

Reply:
It's not necessary to compute the hash of `k` twice, and if `k` is already a valid key, it's not necessary to look up where its value is stored twice. Neither `Dict` nor `OrderedDict` provide an API to do this, however.

Reply:
I guess I should have clarified "contingent on the current API."